core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -89,61 +70,84 @@
         PROHIBITED_VALUE_CHAR_NAMES['\r'] = "<CR>";
     }
 
-    static final int DEFAULT_SIZE_HINT = 16;
-
-    /**
-     * Constant used to seed the hash code generation. Could be anything but this was borrowed from murmur3.
-     */
-    static final int HASH_CODE_SEED = 0xc2b2ae35;
-
-    // XXX(anuraaga): It could be an interesting future optimization if we can use something similar
-    //                to an EnumSet when it's applicable, with just one each of commonly known header names.
-    //                It should be very common.
-    @VisibleForTesting
-    final HeaderEntry[] entries;
-    private final byte hashMask;
-
-    private final HeaderEntry head;
-    private HeaderEntry firstNonPseudo;
-
-    int size;
     private boolean endOfStream;
 
     HttpHeadersBase(int sizeHint) {
-        // Enforce a bound of [2, 128] because hashMask is a byte. The max possible value of hashMask is
-        // one less than the length of this array, and we want the mask to be > 0.
-        entries = new HeaderEntry[findNextPositivePowerOfTwo(max(2, min(sizeHint, 128)))];
-        hashMask = (byte) (entries.length - 1);
-        head = firstNonPseudo = new HeaderEntry();
+        super(sizeHint);
     }
 
     /**
      * Creates a shallow or deep copy of the specified {@link HttpHeadersBase}.
      */
-    HttpHeadersBase(HttpHeadersBase headers, boolean shallowCopy) {
-        hashMask = headers.hashMask;
-        endOfStream = headers.endOfStream;
-
-        if (shallowCopy) {
-            entries = headers.entries;
-            head = headers.head;
-            firstNonPseudo = headers.firstNonPseudo;
-            size = headers.size;
-        } else {
-            entries = new HeaderEntry[headers.entries.length];
-            head = firstNonPseudo = new HeaderEntry();
-            final boolean succeeded = addFast(headers);
-            assert succeeded;
-        }
+    HttpHeadersBase(HttpHeadersBase parent, boolean shallowCopy) {
+        super(parent, shallowCopy);
+        endOfStream = parent.endOfStream;
     }
 
     /**
      * Creates a deep copy of the specified {@link HttpHeaderGetters}.
      */
-    HttpHeadersBase(HttpHeaderGetters headers) {
-        this(headers.size());
-        assert !(headers instanceof HttpHeadersBase);
-        addSlow(headers);
+    HttpHeadersBase(HttpHeaderGetters parent) {
+        super(parent);
+        assert !(parent instanceof HttpHeadersBase);
+        endOfStream = parent.isEndOfStream();
+    }
+
+    @Override
+    int hashName(CharSequence name) {
+        return AsciiString.hashCode(name);
+    }
+
+    @Override
+    boolean nameEquals(@Nullable AsciiString a, @Nullable CharSequence b) {
+        return a != null && (a == b || a.contentEqualsIgnoreCase(b));
```

Comment:
Think we can remove `a == b` now since this should have been released a while ago

https://github.com/netty/netty/pull/9130

Reply:
Note for other comment - here if `a == null` it always returns `false` even if `b == null`

Reply:
Good point - Made `nameEquals` accept nonnull parameters and handled null in the super class.